### ISSUE TITLE
[Bug] Segfault in new sampling pipeline when Debug mode is on

### ISSUE THREAD

**Issue by BarclayII** [dmlc/dgl#3755]
After merging #3665 the following code crashes in when DGL is built with DEBUG mode:
```python
import torch
import dgl
from ogb.nodeproppred import DglNodePropPredDataset

dataset = DglNodePropPredDataset('ogbn-products')
graph, labels = dataset[0]
graph.ndata['label'] = labels.squeeze()
graph.create_formats_()

sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1)
dataloader = dgl.dataloading.NodeDataLoader(
        graph, torch.arange(graph.num_nodes()), sampler, device='cuda',
        batch_size=4096, shuffle=False, drop_last=False, num_workers=1,
        pin_memory=False, use_prefetch_thread=False, use_alternate_streams=False)
for it, _ in enumerate(dataloader):
    print(it)
```
The worker crashes with segfault with the following backtrace:
```
#0  raise (sig=<optimized out>) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f09c4cc4408 in handler_SIGSEGV(int, siginfo_t*, void*) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#2  <signal handler called>
#3  _int_malloc (av=av@entry=0x7f09c7eafc40 <main_arena>, bytes=bytes@entry=5765296) at malloc.c:3924
#4  0x00007f09c7b5967b in _int_memalign (av=0x7f09c7eafc40 <main_arena>, alignment=64, bytes=<optimized out>) at malloc.c:4683
#5  0x00007f09c7b5f0fa in _mid_memalign (address=<optimized out>, bytes=5765192, alignment=<optimized out>) at malloc.c:3324
#6  __posix_memalign (memptr=0x7ffccfe12fd0, alignment=<optimized out>, size=5765192) at malloc.c:5361
#7  0x00007f0967bffcde in c10::alloc_cpu(unsigned long) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libc10.so
#8  0x00007f0967c01452 in c10::DefaultCPUAllocator::allocate(unsigned long) const () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libc10.so
#9  0x00007f09aeabf7e9 in at::detail::empty_generic(c10::ArrayRef<long>, c10::Allocator*, c10::DispatchKey, c10::ScalarType, c10::Device, c10::optional<c10::MemoryFormat>) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#10 0x00007f09aeac0547 in at::detail::empty_cpu(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) ()
   from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#11 0x00007f09aef02af9 in at::native::empty_cpu(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) ()
   from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#12 0x00007f09af5e617a in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>), &at::(anonymous namespace)::(anonymous namespace)::wrapper_memory_format_empty_memory_format>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat> > >, at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) ()
   from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#13 0x00007f09af5c462e in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>), &at::(anonymous namespace)::empty_memory_format>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat> > >, at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#14 0x00007f09af2cca90 in at::_ops::empty_memory_format::call(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) ()
   from /home/ubuntu/miniconda3/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#15 0x00007f07b809d92a in at::empty(c10::ArrayRef<long>, c10::TensorOptions, c10::optional<c10::MemoryFormat>) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/dgl-0.8-py3.8-linux-x86_64.egg/dgl/tensoradapter/pytorch/libtensoradapter_pytorch_1.10.0.so
#16 0x00007f07b809f033 in torch::empty(c10::ArrayRef<long>, c10::TensorOptions, c10::optional<c10::MemoryFormat>) () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/dgl-0.8-py3.8-linux-x86_64.egg/dgl/tensoradapter/pytorch/libtensoradapter_pytorch_1.10.0.so
#17 0x00007f07b809ba3a in TAempty () from /home/ubuntu/miniconda3/lib/python3.8/site-packages/dgl-0.8-py3.8-linux-x86_64.egg/dgl/tensoradapter/pytorch/libtensoradapter_pytorch_1.10.0.so
#18 0x00007f07cc6ad0e3 in dgl::runtime::TensorDispatcher::Empty (this=0x7f07cef559e0 <dgl::runtime::TensorDispatcher::Global()::inst>, shape=std::vector of length 1, capacity 1 = {...}, dtype=..., ctx=...) at /home/ubuntu/dgl/include/dgl/runtime/tensordispatch.h:77
#19 0x00007f07cc6aa9bf in dgl::runtime::NDArray::Empty (shape=std::vector of length 1, capacity 1 = {...}, dtype=..., ctx=...) at /home/ubuntu/dgl/src/runtime/ndarray.cc:214
#20 0x00007f07cc4e6f4a in dgl::aten::impl::CSRToCOO<(DLDeviceType)1, long> (csr=...) at /home/ubuntu/dgl/src/array/cpu/spmat_op_impl_csr.cc:291
#21 0x00007f07cc2537a7 in dgl::aten::CSRToCOO (csr=..., data_as_order=false) at /home/ubuntu/dgl/src/array/array.cc:469
#22 0x00007f07cc8760ca in dgl::UnitGraph::CSR::OutEdges (this=0x55609089d9f0, etype=0, vids=...) at /home/ubuntu/dgl/src/graph/unit_graph.cc:693
#23 0x00007f07cc8631c1 in dgl::UnitGraph::InEdges (this=0x556090499190, etype=0, vids=...) at /home/ubuntu/dgl/src/graph/unit_graph.cc:1017
#24 0x00007f07cc72250c in dgl::HeteroGraph::InEdges (this=0x556090896650, etype=0, vids=...) at /home/ubuntu/dgl/src/graph/./heterograph.h:135
#25 0x00007f07cc7b9691 in dgl::sampling::SampleNeighbors (hg=std::shared_ptr<dgl::BaseHeteroGraph> (use count 3, weak count 0) = {...}, nodes=std::vector of length 1, capacity 1 = {...}, fanouts=std::vector of length 1, capacity 1 = {...}, dir=dgl::EdgeDir::kIn,
    prob=std::vector of length 1, capacity 1 = {...}, exclude_edges=std::vector of length 0, capacity 0, replace=false) at /home/ubuntu/dgl/src/graph/sampling/neighbor/neighbor.cc:101
...
```
Setting `BUILD_TORCH=OFF` still segfaults in the same call in 19th frame (`NDArray::Empty`).

**Comment by VoVAllen** [comment]
does this happen for release mode？

**Comment by BarclayII** [comment]
@VoVAllen It happens occasionally.  I think the segfaults @Rhett-Ying and @yaox12 observed should be caused by the same issue.

**Comment by VoVAllen** [comment]
Did this setting use UVA?

**Comment by BarclayII** [comment]
`-fsanitizer=address` reports a heap-use-after-free error in `sample_neighbors`
```
==40824==ERROR: AddressSanitizer: heap-use-after-free on address 0x7f724f5b0800 at pc 0x7f748d717bcb bp 0x7ffdd8adced0 sp 0x7ffdd8adcec0
READ of size 8 at 0x7f724f5b0800 thread T0
==40700==AddressSanitizer CHECK failed: ../../../../src/libsanitizer/asan/asan_interceptors.cc:384 "((__interception::real___cxa_throw)) != (0)" (0x0, 0x0)
    #0 0x7f748d717bca in dgl::aten::CSRMatrix dgl::aten::impl::CSRSliceRows<(DLDeviceType)1, long>(dgl::aten::CSRMatrix, dgl::runtime::NDArray) /home/ubuntu/dgl/src/array/cpu/spmat_op_impl_csr.cc:400
    #1 0x7f748d1d25bf in dgl::aten::CSRSliceRows(dgl::aten::CSRMatrix, dgl::runtime::NDArray) /home/ubuntu/dgl/src/array/array.cc:496
    #2 0x7f748dec7cac in dgl::UnitGraph::CSR::OutEdges(unsigned long, dgl::runtime::NDArray) const /home/ubuntu/dgl/src/graph/unit_graph.cc:680
    #3 0x7f748de9b387 in dgl::UnitGraph::InEdges(unsigned long, dgl::runtime::NDArray) const /home/ubuntu/dgl/src/graph/unit_graph.cc:993
    #4 0x7f748dbc6a82 in dgl::HeteroGraph::InEdges(unsigned long, dgl::runtime::NDArray) const /home/ubuntu/dgl/src/graph/./heterograph.h:135
    #5 0x7f748dd1adf1 in dgl::sampling::SampleNeighbors(std::shared_ptr<dgl::BaseHeteroGraph>, std::vector<dgl::runtime::NDArray, std::allocator<dgl::runtime::NDArray> > const&, std::vector<long, std::allocator<long> > const&, dgl::EdgeDir, std::vector<dgl::runtime::NDArray, std::allocator<dgl::runtime::NDArray> > const&, std::vector<dgl::runtime::NDArray, std::allocator<dgl::runtime::NDArray> > const&, bool) /home/ubuntu/dgl/src/graph/sampling/neighbor/neighbor.cc:101
    #6 0x7f748dd2a69d in operator() /home/ubuntu/dgl/src/graph/sampling/neighbor/neighbor.cc:454
    #7 0x7f748dd2d278 in _M_invoke /usr/include/c++/7/bits/std_function.h:316
    #8 0x7f748da8337f in std::function<void (dgl::runtime::DGLArgs, dgl::runtime::DGLRetValue*)>::operator()(dgl::runtime::DGLArgs, dgl::runtime::DGLRetValue*) const /usr/include/c++/7/bits/std_function.h:706
    #9 0x7f748da80917 in dgl::runtime::PackedFunc::CallPacked(dgl::runtime::DGLArgs, dgl::runtime::DGLRetValue*) const /home/ubuntu/dgl/include/dgl/runtime/packed_func.h:817
    #10 0x7f748da7a5f2 in DGLFuncCall /home/ubuntu/dgl/src/runtime/c_runtime_api.cc:264
```

So the seed array has been

**Comment by BarclayII** [comment]
No.  Even CPU doesn't work.

Since the stackframe says that `rows` is freed, I suspect the problematic code is in how I shuffle stuff: https://github.com/dmlc/dgl/blob/7b9afbfa28e9441fd3bfbe4e3c24eb62816cb1a7/python/dgl/dataloading/dataloader.py#L113-L117

Modifying in-place is unlikely to be correct.  I think I should create a shared memory array and broadcast the array handle to the workers instead.

**Comment by VoVAllen** [comment]
Actually I cannot reproduce the crash on my side, with `cmake -DUSE_CUDA=ON -DCMAKE_BUILD_TYPE=DEBUG ..` and the script you posted...

**Comment by BarclayII** [comment]
This is the CMake command I used:
```
cmake -DCMAKE_SHARED_LINKER_FLAGS='-fsanitize=address -static-libasan' -DCMAKE_C_FLAGS='-static-libasan -fsanitize=address' -DCMAKE_CXX_FLAGS='-static-libasan -fsanitize=address' -DUSE_OPENMP=OFF -DUSE_CUDA=ON -DBUILD_TORCH=OFF -DCMAKE_BUILD_TYPE=Debug -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-10.2 ..
```
Did the sanitizer complain about heap-use-after-free in your case?

**Comment by BarclayII** [comment]
Adding `clone()` to the iterator's returns magically solves the problem.  Don't know why so I guess I'll leave this issue open?
